### PR TITLE
deprecated. integerate poseidon merkletree

### PR DESCRIPTION
**Pull request by lightsing**
Status:

1. all unittest  passed ( except those disable for a legit reason, explain below)
2. USDC/UniswapV3 e2e test suites passed
3. we use golang build flags to control using old trie or our new zktrie. (TODO: make sure default mode is old trie)

Behavior change:
1. state/snapshot disabled ( I even doubt whether it is needed since we trace every opcode)
2. fasy sync disabled. Snap sync is a better & modern option.
3. state gc disabled. Always archive node

TODO(before merge)

- [ ] 1. merge https://github.com/scroll-tech/go-ethereum/pull/79
- [ ] 2. (WIP) performance test of native and erc20 transfer
- [ ] 3. rename "smt" to "zktrie"; remove useless circom stuff & code clean ( like removing some sanity check)
- [ ] 4. basic docs/comments describing changes/design

TODO(long term, nice to have):
- [ ] 1. fix snap & snap sync
- [ ] 2. use preimage cache? 
- [ ] 3. recover metric codes..
- [ ] 4. gc / prune / freeze
- [ ] 5. using stronger type, (like replacing bigint with finitefield), to prevent bugs

### CODE REVIEW

**Comment by ChuhanJin** [comment]
Can one of the admins verify this patch?

**Comment by lispc** [comment]
it is a big PR. so maybe reviewers can have a quick & rough look first.

File: trie/secure_trie.go [hunk 1]
```diff
@@ -36,32 +42,32 @@ import (
...
-	if db == nil {
-		panic("trie.NewSecure called without a database")
+	tree *MerkleTree
+}
+
+var mtDbOnce sync.Once
+var mtDb db.Storage
+
+// NewSecure creates a trie
+func NewSecure(root common.Hash, _ *Database) (*SecureTrie, error) {
```
Comment (mask-pp) [inline trie/secure_trie.go]: Why not use db by parameter?

File: core/types/state_account.go [hunk 2]
```diff
@@ -30,3 +39,73 @@ type StateAccount struct {
 	Root     common.Hash // merkle root of the storage trie
 	CodeHash []byte
 }
+
+// Hash of StateAccount
+// AccountHash = Hash(
```
Comment (mask-pp) [inline core/types/state_account.go:45]: That should be deleted？

File: go.mod [hunk 3]
```diff
@@ -12,13 +12,14 @@ require (
...
 	github.com/cloudflare/cloudflare-go v0.14.0
+	github.com/cockroachdb/pebble v0.0.0-20220127033822-015f5e3806d3
 	github.com/consensys/gnark-crypto v0.4.1-0.20210426202927-39ac3d4b3f1f
 	github.com/davecgh/go-spew v1.1.1
 	github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea
 	github.com/docker/docker v1.4.2-0.20180625184442-8e610b2b55bf
 	github.com/dop251/goja v0.0.0-20211011172007-d99e4b8cbf48
 	github.com/edsrzf/mmap-go v1.0.0
-	github.com/ethereum/go-ethereum v1.10.13
+	github.com/ethereum/go-ethereum v1.9.12
```
Comment (mask-pp) [inline go.mod]: L2geth based on geth@v1.10.13, should be restored.

File: trie/database.go [hunk 4]
```diff
@@ -511,27 +511,28 @@ func (db *Database) reference(child common.Hash, parent common.Hash) {
...
-	db.gcsize += storage - db.dirtiesSize
-	db.gctime += time.Since(start)
-
-	memcacheGCTimeTimer.Update(time.Since(start))
-	memcacheGCSizeMeter.Mark(int64(storage - db.dirtiesSize))
-	memcacheGCNodesMeter.Mark(int64(nodes - len(db.dirties)))
-
-	log.Debug("Dereferenced trie from memory database", "nodes", nodes-len(db.dirties), "size", storage-db.dirtiesSize, "time", time.Since(start),
-		"gcnodes", db.gcnodes, "gcsize", db.gcsize, "gctime", db.gctime, "livenodes", len(db.dirties), "livesize", db.dirtiesSize)
+	//// Sanity check to ensure that the meta-root is not removed
```
Comment (mask-pp) [inline trie/database.go]: Why not delete useless code?

Comment (mask-pp) [inline trie/database.go:514]: Same as above

File: trie/trie.go [hunk 5]
```diff
@@ -534,39 +532,39 @@ func (t *Trie) Commit(onleaf LeafCallback) (common.Hash, int, error) {
...
-		// channel here.
-		close(h.leafCh)
-		wg.Wait()
-	}
-	if err != nil {
-		return common.Hash{}, 0, err
-	}
-	t.root = newRoot
-	return rootHash, committed, nil
+	//h := newCommitter()
```
Comment (mask-pp) [inline trie/trie.go]: Why not delete that?

File: trie/db/db.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
+package db
```
Comment (mask-pp) [inline trie/db/db.go:1]: We'd better reference ethdb.

File: trie/secure_trie.go (hunk 1 above)
Comment (lightsing) [reply trie/secure_trie.go]: 1. iden3 mekletree needs a raw leveldb.
2. create at leveldb at here is a temporary workaround

File: core/types/state_account.go (hunk 2 above)
Comment (lightsing) [reply core/types/state_account.go:45]: Why? This comment describes how we calculate the AccountHash

Comment (mask-pp) [reply core/types/state_account.go:45]: Purely Use sentence maybe better, that like useless code.

File: core/types/smt/hash.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	"math/big"
+	"strings"
+)
+
+const numCharPrint = 8
+
+// ElemBytesLen is the length of the Hash byte array
+const ElemBytesLen = 32
+
+var HashZero = Hash{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
```
Comment (mask-pp) [inline core/types/smt/hash.go]: Use `var HashZero = Hash{}` is enough.

File: core/types/smt/byte32.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+import (
+	"github.com/iden3/go-iden3-crypto/poseidon"
+	"math/big"
+)
+
+type Byte32 [32]byte
+
+func (b *Byte32) Hash() (*big.Int, error) {
+	first16 := new(big.Int).SetBytes(b[0:16])
```
Comment (mask-pp) [inline core/types/smt/byte32.go:14]: Just use `new(big.Int).SetBytes(b)` is enough?

File: core/types/smt/hash.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		return nil, fmt.Errorf("Expected 32 bytes, found %d bytes", len(b))
+	}
+	var h Hash
+	copy(h[:], SwapEndianness(b))
+	return &h, nil
+}
+
+// NewHashFromHex returns a *Hash representation of the given hex string
+func NewHashFromHex(h string) (*Hash, error) {
+	h = strings.TrimPrefix(h, "0x")
```
Comment (mask-pp) [inline core/types/smt/hash.go]: Use `common.FromHex(h)` is enough.

File: common/types.go [hunk 10]
```diff
@@ -228,6 +228,12 @@ func IsHexAddress(s string) bool {
 // Bytes gets the string representation of the underlying address.
 func (a Address) Bytes() []byte { return a[:] }
 
+func (a Address) Bytes32() []byte {
```
Comment (mask-pp) [inline common/types.go:232]: Use `NewByte32FromBytesPaddingZero()` is enough?

File: trie/secure_trie.go [hunk 11]
```diff
@@ -123,12 +137,12 @@ func (t *SecureTrie) Update(key, value []byte) {
 //
 // If a node was not found in the database, a MissingNodeError is returned.
 func (t *SecureTrie) TryUpdate(key, value []byte) error {
-	hk := t.hashKey(key)
-	err := t.trie.TryUpdate(hk, value)
+	kPreimage := NewByte32FromBytesPaddingZero(key)
+	vPreimage := NewByte32FromBytesPaddingZero(value)
```
Comment (mask-pp) [inline trie/secure_trie.go]: It's right to use `NewByte32FromBytesPaddingZero(value)`?

File: core/types/smt/hash.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+func NewHashFromString(s string) (*Hash, error) {
+	bi, ok := new(big.Int).SetString(s, 10)
+	if !ok {
+		return nil, fmt.Errorf("Can not parse string to Hash")
+	}
+	return NewHashFromBigInt(bi), nil
+}
+
+// SwapEndianness swaps the order of the bytes in the slice.
+func SwapEndianness(b []byte) []byte {
```
Comment (mask-pp) [inline core/types/smt/hash.go]: Is Endianness change func?
It should for the word(uint64,i32...), but not for []byte.
Please reference `binary.BigEndian`.

File: core/types/smt/byte32.go (hunk 8 above)
Comment (lightsing) [reply core/types/smt/byte32.go:14]: no, need to seperate into two parts

**Comment by lispc** [comment]
1. error of TryUpdate
2. docs of changes

**Comment by lispc** [comment]
"I feel that there may still be a logical problem with the call relationship in the AddBalance place.
Suppose Miner is a new Account and currently has no record in World State/SMT. So now the call relationship for assigning values to this account is AddBalance -> GetOrNewStateObject -> getStateObject -> getDeletedStateObject -> TryGet -> GetLeafNodeByWord.
The problem is in the two functions TryGet and GetLeafNodeByWord. In the current code, if the leaf node is not found, then the functions return nil, err. Then for a new Account, these two functions return err. This err causes the stateDB to Panic off when calling commit (in stated.go). This may be missing a logic to create an Account in SMT."          

I am a bit confused that according to this comment, current code should not run smoothly to create new account(transfer to new addr, deploy contract..). ...

It seems in original geth, (nil, nil) is [returned](https://github.com/ethereum/go-ethereum/blob/37f9d25ba027356457953eab5f181c98b46e9988/trie/trie.go#L135)  when `TryGet(non_exist_key)` is called, but in our trie, (nil, err) is returned?

**Comment by hsyodyssey** [comment]
``> "I feel that there may still be a logical problem with the call relationship in the AddBalance place. Suppose Miner is a new Account and currently has no record in World State/SMT. So now the call relationship for assigning values to this account is AddBalance -> GetOrNewStateObject -> getStateObject -> getDeletedStateObject -> TryGet -> GetLeafNodeByWord. The problem is in the two functions TryGet and GetLeafNodeByWord. In the current code, if the leaf node is not found, then the functions return nil, err. Then for a new Account, these two functions return err. This err causes the stateDB to Panic off when calling commit (in stated.go). This may be missing a logic to create an Account in SMT."
> 
> I am a bit confused that according to this comment, current code should not run smoothly to create new account(transfer to new addr, deploy contract..). ...

Oh my fault, exactly now statedb is not panic, but still print error. 
```
WARN [03-04|23:30:39.580] Served eth_getBalance                    conn=127.0.0.1:52193 reqid=42563 t="141.562µs" err="getDeleteStateObject (092c8103de357eda16da05229542c62a657b9f0d) error: Key not found in the MerkleTree"
Key not found in the MerkleTree
```
My output here will keep printing this error because of the following code.
```
func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
	if s.dbErr != nil {
		return common.Hash{}, fmt.Errorf("commit aborted due to earlier error: %v", s.dbErr)
	}
```

**Comment by hsyodyssey** [comment]
> ``> "I feel that there may still be a logical problem with the call relationship in the AddBalance place. Suppose Miner is a new Account and currently has no record in World State/SMT. So now the call relationship for assigning values to this account is AddBalance -> GetOrNewStateObject -> getStateObject -> getDeletedStateObject -> TryGet -> GetLeafNodeByWord. The problem is in the two functions TryGet and GetLeafNodeByWord. In the current code, if the leaf node is not found, then the functions return nil, err. Then for a new Account, these two functions return err. This err causes the stateDB to Panic off when calling commit (in stated.go). This may be missing a logic to create an Account in SMT."
> 
> > I am a bit confused that according to this comment, current code should not run smoothly to create new account(transfer to new addr, deploy contract..). ...
> 
> Oh my fault, exactly now statedb is not panic, but still print error.
> 
> ```
> WARN [03-04|23:30:39.580] Served eth_getBalance                    conn=127.0.0.1:52193 reqid=42563 t="141.562µs" err="getDeleteStateObject (092c8103de357eda16da05229542c62a657b9f0d) error: Key not found in the MerkleTree"
> Key not found in the MerkleTree
> ```
> 
> My output here will keep printing this error because of the following code.
> 
> ```
> func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
> 	if s.dbErr != nil {
> 		return common.Hash{}, fmt.Errorf("commit aborted due to earlier error: %v", s.dbErr)
> 	}
> ```

When I test the current codebase by making a new geth locally, I encounter the following issues. This shall be related to this issue mentioned here🙏.

image

**Comment by hsyodyssey** [comment]
Hi guys, I ran the current code and so far I only found that at the time of genesis block generation (ToBlock function), the commit function was not fully executed and returned early due to the existence of an err in the Statedb instance. However, we can see in the genesis block code that it does not handle the error returned by statedb when calling the statedb.commit function, so even though statedb.commit did not execute and returned an error directly, it still continued to execute.

```
func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
….
statedb.Commit(false)
…
}
```

Another statedb.commit function call is located in the blockchain/writeBlockWithState function. In this function, it reacts to the failures of a commit in some way. But it doesn't seem to have the error here yet.

```
func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.Receipt, logs []*types.Log, blockResult *types.BlockResult, state *state.StateDB, emitHeadEvent bool) (status WriteStatus, err error) {
…
root, err := state.Commit(bc.chainConfig.IsEIP158(block.Number()))
if err != nil {
	log.Error("State Commit Error", err)
	return NonStatTy, err
}
…
}
```

This should be the reason why the current code works fine.

I think we need to do more tests to verify that. Not quite sure if there are any follow up effects of the failure of the commit function there in genesis generation?

**Comment by hsyodyssey** [comment]
The program failed when I called the function creatToken in the following contract. According to the logs, this issue may have happened in StorageTrie function (cpy.updateTrie(s.db)).

Here are the related logs.

```
github.com/scroll-tech/go-ethereum/core/types/smt.NewByte32FromBytesPaddingZero(...)
	github.com/scroll-tech/go-ethereum/core/types/smt/byte32.go:22
github.com/scroll-tech/go-ethereum/trie.(*SecureTrie).TryUpdate(0xc012650920, {0xc0011d8a40, 0x20, 0x30}, {0xc01c6ec210, 0x21, 0xc001134160})
	github.com/scroll-tech/go-ethereum/trie/secure_trie.go:147 +0x1b1
github.com/scroll-tech/go-ethereum/core/state.(*stateObject).updateTrie(0xc00112d2c0, {0x5576430, 0xc000592318})
	github.com/scroll-tech/go-ethereum/core/state/state_object.go:353 +0x98d
github.com/scroll-tech/go-ethereum/core/state.(*StateDB).StorageTrie(0xc00130e000, {0x4e, 0x35, 0x1c, 0x25, 0x92, 0x43, 0x9b, 0x64, 0x6d, ...})
	github.com/scroll-tech/go-ethereum/core/state/statedb.go:362 +0x6d
github.com/scroll-tech/go-ethereum/core/state.(*StateDB).GetStorageProof(0x0, {0x4e, 0x35, 0x1c, 0x25, 0x92, 0x43, 0x9b, 0x64, 0x6d, ...}, ...)
	github.com/scroll-tech/go-ethereum/core/state/statedb.go:332 +0x58
github.com/scroll-tech/go-ethereum/core/vm.traceStorageProof(0xc0005c4150, 0xc011921b60, 0xc011a40a80)
	github.com/scroll-tech/go-ethereum/core/vm/logger_trace.go:76 +0x12f
github.com/scroll-tech/go-ethereum/core/vm.(*StructLogger).CaptureStateAfter(0xc0005c4150, 0xbe, 0x55, 0x2dfd2, 0x4e20, 0xc011921b60, {0x0, 0x0, 0x251c354e00000000}, 0x2, ...)
	github.com/scroll-tech/go-ethereum/core/vm/logger.go:241 +0x1ba
github.com/scroll-tech/go-ethereum/core/vm.(*EVMInterpreter).Run(0xc001114000, 0xc012084900, {0x0, 0x0, 0x0}, 0x0)
	github.com/scroll-tech/go-ethereum/core/vm/interpreter.go:269 +0xbf0
github.com/scroll-tech/go-ethereum/core/vm.(*EVM).create(0xc001104000, {0x554b000, 0xc012084840}, 0xc001134b98, 0x3836f, 0x6552f18, {0x4e, 0x35, 0x1c, 0x25, ...}, ...)
	github.com/scroll-tech/go-ethereum/core/vm/evm.go:469 +0x7f6
github.com/scroll-tech/go-ethereum/core/vm.(*EVM).Create(0xc001104000, {0x554b000, 0xc012084840}, {0xc001126000, 0x438, 0x438}, 0x0, 0x5)
	github.com/scroll-tech/go-ethereum/core/vm/evm.go:517 +0x1e5
github.com/scroll-tech/go-ethereum/core/vm.opCreate(0x1, 0xc001114000, 0xc011921698)
	github.com/scroll-tech/go-ethereum/core/vm/instructions.go:585 +0x3dd
github.com/scroll-tech/go-ethereum/core/vm.(*EVMInterpreter).Run(0xc001114000, 0xc012084840, {0xc000549260, 0x24, 0x24}, 0x0)
	github.com/scroll-tech/go-ethereum/core/vm/interpreter.go:261 +0xad3
github.com/scroll-tech/go-ethereum/core/vm.(*EVM).Call(0xc001104000, {0x554e000, 0xc004bc8360}, {0x9d, 0xc3, 0x4b, 0x69, 0x19, 0xe2, 0xe0, ...}, ...)
	github.com/scroll-tech/go-ethereum/core/vm/evm.go:231 +0xc34
github.com/scroll-tech/go-ethereum/core.(*StateTransition).TransitionDb(0xc001094800)
	github.com/scroll-tech/go-ethereum/core/state_transition.go:325 +0x7fc
github.com/scroll-tech/go-ethereum/core.ApplyMessage(0x558ca10, {0x558ca10, 0xc00055efc0}, 0xc004bc8318)
	github.com/scroll-tech/go-ethereum/core/state_transition.go:181 +0x2a
github.com/scroll-tech/go-ethereum/core.applyTransaction({0xc004bc8318, {0x78, 0xb2, 0x84, 0xc3, 0x31, 0x21, 0x19, 0xfb, 0xc8, ...}, ...}, ...)
	github.com/scroll-tech/go-ethereum/core/state_processor.go:101 +0x1cb
github.com/scroll-tech/go-ethereum/core.ApplyTransaction(_, {_, _}, _, _, _, _, _, _, {0x1, ...})
	github.com/scroll-tech/go-ethereum/core/state_processor.go:157 +0x368
github.com/scroll-tech/go-ethereum/miner.(*worker).commitTransaction(0xc000448480, 0xc013c484e0, {0x78, 0xb2, 0x84, 0xc3, 0x31, 0x21, 0x19, 0xfb, ...})
	github.com/scroll-tech/go-ethereum/miner/worker.go:788 +0x270
github.com/scroll-tech/go-ethereum/miner.(*worker).commitTransactions(0xc000448480, 0xc011a40480, {0x78, 0xb2, 0x84, 0xc3, 0x31, 0x21, 0x19, 0xfb, ...}, ...)
	github.com/scroll-tech/go-ethereum/miner/worker.go:865 +0x51b
github.com/scroll-tech/go-ethereum/miner.(*worker).commitNewWork(0xc000448480, 0xc00065fc6c, 0x1, 0x62237798)
	github.com/scroll-tech/go-ethereum/miner/worker.go:1037 +0xf25
github.com/scroll-tech/go-ethereum/miner.(*worker).mainLoop(0xc000448480)
	github.com/scroll-tech/go-ethereum/miner/worker.go:468 +0x67d
created by github.com/scroll-tech/go-ethereum/miner.newWorker
	github.com/scroll-tech/go-ethereum/miner/worker.go:235 +0x725
```

**Comment by lispc** [comment]
As for the failure in createToken contract call, 
(1) i found all contract call in this repo is run in trace mode, and when tracing, storage proof is generated along with every sload/sstore.  Since trie.Prove is under implementation by @noel2004 , the failure is expected.
(2) we should remove rlp encoding here <https://github.com/scroll-tech/go-ethereum/blob/2040f8c78170b0c0d95b95361ded7a5e4848ca2e/core/state/state_object.go#L352> @lightsing

File: core/types/smt/hash.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	return new(big.Int).SetBytes(ReverseByteOrder(h[:]))
+}
+
+// Bytes returns the []byte representation of the *Hash, which always is 32
+// bytes length.
+func (h *Hash) Bytes() []byte {
+	bi := new(big.Int).SetBytes(h[:]).Bytes()
+	b := [32]byte{}
+	copy(b[:], ReverseByteOrder(bi[:]))
```
Comment (mask-pp) [inline core/types/smt/hash.go:66]: Can return `ReverseByteOrder(bi[:])` directly.

**Comment by lispc** [comment]
TODO:    https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/state/statedb.go#L478. (remove RLP?)

**Comment by lightsing** [comment]
## performance changes

**THE PREVIOUS BENCH USES WRONG METHOD.**

### Result

Following tests are running on same Ubuntu server. tps is estimated with normal dist.
Bench using integration-test/bench branch
geth start cmd: https://github.com/scroll-tech/integration-test/blob/bench/scripts/start.sh#L11

### Native transaction
- zkrollup estimated with 912.425 tps (19 valid samples)
- zkrollup-smt estimated with 369.869 tps (-59.46%) (13 valid samples)

### erc20 transaction
- zkrollup estimated with 106.47 tps (8 valid samples)
- zkrollup-smt estimated with 118.56 tps (15 valid samples)
no significant performance difference.

**Review by icemelon — COMMENTED** [review]
Is it possible that we could add a config to switch between `SecureTrie` and `SecureBinaryTrie`? But it seems pretty hard from the first glance...

**Comment by lispc** [comment]
> Is it possible that we could add a config to switch between `SecureTrie` and `SecureBinaryTrie`? But it seems pretty hard from the first glance...

I think not very hard... I guess one week X developer may be ok for this...

**Comment by lispc** [comment]
> ## performance changes
> **THE PREVIOUS BENCH USES WRONG METHOD.**
> 
> ### Result
> Following tests are running on same Ubuntu server. tps is estimated with normal dist.
> 
> ### Native transaction
> * zkrollup estimated with 912.425 tps (19 valid samples)
> * zkrollup-smt estimated with 369.869 tps (-59.46%) (13 valid samples)
> 
> ### erc20 transaction
> * zkrollup estimated with 106.47 tps (8 valid samples)
> * zkrollup-smt estimated with 118.56 tps (15 valid samples)
>   no significant performance difference.

ok. Also add the geth launch cmd in the comments here. And git commits for two branch

**Comment by icemelon** [comment]
@lispc @lightsing 
If it's not a significant amount of work, I think it's worth adding an option to switch between `SecureTrie` and `SecureBinaryTrie` so in the future it won't be a large effort if we want to switch back.

File: core/types/smt/byte32.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	last16 := new(big.Int).SetBytes(b[16:32])
+	hash, err := poseidon.Hash([]*big.Int{first16, last16})
+	if err != nil {
+		return nil, err
+	}
+	return hash, nil
```
Comment (jules) [inline core/types/smt/byte32.go:20]: Can simplify this to one line: 

```golang
return poseidon.Hash([]*big.Int{first16, last16})
```

File: core/types/smt/hash.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+func (h *Hash) Bytes() []byte {
+	buf := make([]byte, 32)
+	bi := new(big.Int).SetBytes(h[:])
+	bi.FillBytes(buf)
+	result := ReverseByteOrder(buf[:])
+
+	return result
+}
+
+func (h *Hash) Bytes2() []byte {
```
Comment (jules) [inline core/types/smt/hash.go:73]: Should this be `Bytes32`?

File: core/types/state_account.go [hunk 16]
```diff
@@ -30,3 +40,76 @@ type StateAccount struct {
...
+//	Hash(nonce, balance),
+//  Hash(
+//	  Root,
```
Comment (jules) [inline core/types/state_account.go:48]: Should we denote this as `Hash(Root),`, as we hash it in the logic below?

File: trie/merkletree.go [hunk 17]
```diff
@@ -0,0 +1,1242 @@
...
+// NewMerkleTreeWithRoot loads a new Merkletree. If in the storage already exists one
+// will open that one, if not, will create a new one.
+func NewMerkleTreeWithRoot(storage db.Storage, root *smt.Hash, maxLevels int) (*MerkleTree, error) {
+	mt := MerkleTree{db: storage, maxLevels: maxLevels, writable: true}
+	mt.rootKey = root
+	if *root != smt.HashZero {
+		_, err := mt.GetNode(mt.rootKey)
+		if err != nil {
+			return nil, err
+		}
```
Comment (jules) [inline trie/merkletree.go:77]: Small nit for style:

```golang
if _, err := mt.GetNode(mt.rootKey); err != nil {
    return nil, err
}
```

File: trie/merkletree.go [hunk 18]
```diff
@@ -0,0 +1,1242 @@
...
+
+// DB returns the MerkleTree.DB()
+func (mt *MerkleTree) DB() db.Storage {
+	return mt.db
+}
+
+// Root returns the MerkleRoot
+func (mt *MerkleTree) Root() *smt.Hash {
+	// FIXME
+	_, err := mt.GetNode(mt.rootKey)
```
Comment (jules) [inline trie/merkletree.go:90]: Ditto

File: trie/merkletree.go [hunk 19]
```diff
@@ -0,0 +1,1242 @@
...
+// MaxLevels returns the MT maximum level
+func (mt *MerkleTree) MaxLevels() int {
+	return mt.maxLevels
+}
+
+// Snapshot returns a read-only copy of the MerkleTree
+func (mt *MerkleTree) Snapshot(rootKey *smt.Hash) (*MerkleTree, error) {
+	mt.RLock()
+	defer mt.RUnlock()
+	_, err := mt.GetNode(rootKey)
```
Comment (jules) [inline trie/merkletree.go:108]: Ditto

File: trie/merkletree.go [hunk 20]
```diff
@@ -0,0 +1,1242 @@
...
+	if err == ErrEntryIndexAlreadyExists {
+		newRootKey, err = mt.updateNode(tx, newNodeLeaf)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+	mt.rootKey = newRootKey
+	err = mt.dbInsert(tx, dbKeyRootNode, DBEntryTypeRoot, mt.rootKey[:])
```
Comment (jules) [inline trie/merkletree.go:153]: Ditto

File: trie/merkletree.go [hunk 21]
```diff
@@ -0,0 +1,1242 @@
...
+func (mt *MerkleTree) AddWord(kPreimage, vPreimage *smt.Byte32) error {
+	k, err := kPreimage.Hash()
+	if err != nil {
+		return err
+	}
+	v, err := vPreimage.Hash()
+	if err != nil {
+		return err
+	}
+	err = mt.Add(k, v, kPreimage, vPreimage[:])
```
Comment (jules) [inline trie/merkletree.go:176]:
```golang
return mt.Add(k, v, kPreimage, vPreimage[:])
```

File: trie/merkletree.go [hunk 22]
```diff
@@ -0,0 +1,1242 @@
...
+
+// AddVarWord
+// Add a Bytes32 k and a var-length value to MerkleTree
+// User Must provide the hash of vPreimage since there is no define for vPreimage not Bytes32
+func (mt *MerkleTree) AddVarWord(kPreimage *smt.Byte32, vHash *big.Int, vPreimage []byte) error {
+	k, err := kPreimage.Hash()
+	if err != nil {
+		return err
+	}
+	err = mt.Add(k, vHash, kPreimage, vPreimage[:])
```
Comment (jules) [inline trie/merkletree.go:188]: Same as above

File: trie/merkletree.go [hunk 23]
```diff
@@ -0,0 +1,1242 @@
...
+	return &mt, nil
+}
+
+// DB returns the MerkleTree.DB()
+func (mt *MerkleTree) DB() db.Storage {
+	return mt.db
+}
+
+// Root returns the MerkleRoot
+func (mt *MerkleTree) Root() *smt.Hash {
```
Comment (jules) [inline trie/merkletree.go:88]: Should this function read-lock the mutex?

**Comment by icemelon** [comment]
should we close this PR? @lispc

**Comment by lispc** [comment]
the smtIterator here is not merged yet. We may pick it back later when we really need it

**Pull request closed without merging**